### ISSUE TITLE
Protected branch

### ISSUE THREAD

**Issue by truongsinh** [gogits/gogs#776]
Reference https://about.gitlab.com/2014/09/29/gitlab-flow/
Reference https://code.google.com/p/gerrit/

Gogs should have a feature of protected branch. Let say `master` is a protected branch, then:
- Nobody can push directly to `master`
- Everybody\* can checkout/fork from `master`
- Everybody\* submit a merge request / code review request for his/her new branch (features/fixes)
- Everybody\* can comment/discuss the diff, preferable line-by-line
- Only those with "Reviewer" or "Supervisor" or "Senior" privilege can vote up/down on a merge request / code review
- A merge request can only be eligible to be merged to master when having at least [a number] 2 of up votes. This parameter should be defined in a config file, environment variable, or runtime argument, so that nobody, even master or supreme admin can bypass code review process by setting it to 1 and self-review their code.

*Everybody is any user with sufficient authorization to read a project `master` and write to their own name space.

**Comment by gittex** [comment]
First gogs would need some ACL mechanism, then its a matter of checking the ACLs everywhere.

See https://gitlab.com/gitlab-org/gitlab-ce/blob/master/doc/permissions/permissions.md

**Comment by Unknwon** [comment]
@gittex thanks your info!

**Comment by srvg** [comment]
+1 here
Branch permissions are a typical corporate feature that will do a lot of good for gogs.

**Comment by truongsinh** [comment]
ping

**Comment by Unknwon** [comment]
Hi @truongsinh , don't know what to say, hmm...

This feature is great and I personally want to have it, too. But it's still in the waiting list. :joy:

**Comment by ohko** [comment]
+1

**Comment by renewsorensen** [comment]
:+1:

**Comment by drsect0r** [comment]
+1

**Comment by roidelapluie** [comment]
:+1:

**Comment by wenerme** [comment]
+1

**Comment by roidelapluie** [comment]
I have seen the card https://trello.com/c/oJWlgzvE I hope that it will be possible to have that setting per branch.

**Comment by mgracer48** [comment]
+100000000000000000000

**Comment by i00** [comment]
I don't like the idea of votes, just want branch permissions... but also want to take it a step further...

For eg (my case).
 If I have a user called Jack I want to be able to give him permissions to publish new branches that start with Jack/ only (so I know exactly what user published a branch at a glance and don't end up with "junk" branches everywhere).
 I want to be the only one who can push to the master (so I can choose what to merge in).

I want to be able to restrict publishes, pushes and pulls (preferably visibility entirely if I don't grant pull permissions).
 All of this should hopefully be able to be done with wildcards to suit my purposes..
eg:
Permissions for jack:
Unspecified permissions (cannot be deleted) - Pull only
Jack/* - Full
Permissions for me:
Unspecified permissions (cannot be deleted) - Full
Permissions for Dean:
Unspecified permissions (cannot be deleted) - Full
master - Pull only

Thanks, for the great work :)

**Comment by tebanep** [comment]
+1

**Comment by sluedecke** [comment]
+1, definitely an important feature!

**Comment by Twisterat57** [comment]
+1, with also the possiblity to make visible as read-only the wiki pages + issues (read + Write) for some teams:
Purpose of having access to wiki pages in read-only is to share internally setup files with non developer teams (ie: device preparators, Commercial team, and why not in the future, a customer). Then gogs would be the "all in one tool" all development team should have! (ok it should be already the: all developer team should use it already XD )

**Comment by pbi-qfs** [comment]
Isn't it possible to implement simple protection (e.g. only allow some users to push to specific branches) via git-hooks on the server?

**Comment by Unknwon** [comment]
@pbi-qfs post-receive hook?

**Comment by 0xbaadf00d** [comment]
Is possible to partial implement this feature without real permission ? onyl by specify a list of branch to protect and a list of user allowed to push on ?

**Comment by Unknwon** [comment]
@0xbaadf00d no it's not possible, what does "real permission" mean?

**Comment by 0xbaadf00d** [comment]
Something like this: https://gitlab.com/gitlab-org/gitlab-ce/blob/master/doc/permissions/permissions.md

**Comment by Unknwon** [comment]
@0xbaadf00d I think this feature will come with that directly.

**Comment by 0xbaadf00d** [comment]
ok, nice

**Comment by mehrad-rafigh** [comment]
+1

**Comment by denji** [comment]
+1, `git push --force` **will destroy this planet!**

**Comment by joelhy** [comment]
+1

**Comment by bhchew** [comment]
+1, nice to see PR between branches are implemented.

Pending this to prevent manual merges between branches outside PR =)

**Comment by renewsorensen** [comment]
+1

**Comment by rudineirk** [comment]
+1

**Comment by mrkeuz** [comment]
+1

**Comment by FyinInc** [comment]
+1

**Comment by heber-minetto** [comment]
+1

**Comment by Jofkos** [comment]
+1

**Comment by hrobertson** [comment]
I'm guessing :+1:ing the OP is preferred to all these +1 comments?

**Comment by LucksBoy** [comment]
i need this feature

**Comment by relipa** [comment]
So do I

**Comment by b-connect** [comment]
+1

**Comment by Rhys** [comment]
I would also appreciate it.

**Comment by YamiOdymel** [comment]
Still +1

**Comment by lnalex** [comment]
Is someone working on it ?

**Comment by congeal** [comment]
+1

**Comment by ptman** [comment]
I think protected branches / branch permissions can be implemented by itself without having to provide much integration with the code review side initially.

**Comment by lnalex** [comment]
@ptman This is not the same need, of course if you are code reviewing your project you have to protect branches but you also need to make comments on the submitted code: both are different feature.

**Comment by hrobertson** [comment]
@ptman @lnalex As has been said, 'Protected Branches' and 'Line-by-line code comments on the PR' are separate features. Neither _requires_ the other.

If you read the first post of this feature request, it is asking for Protected Branches.
Line-by-line comments are only mentioned as a nice to have:

> - Everybody* can comment/discuss the diff, preferable line-by-line

The existing Pull Requests feature enables commenting, but not line-by-line.

I agree that the ability to add line-by-line comments on a pull request would be a useful feature, but that should be raised as a separate feature request.

Additionally, some people have requested a more comprehensive permissions system. Again, that is a separate feature request: https://github.com/gogits/gogs/issues/1743

**Comment by freshcn** [comment]
+1

**Comment by strk** [comment]
It would be worth for the approval system to be driven by external/arbitrary plugins, using an approval API.
The lgtm project does this to drive approvals for github protected branches, for example:
https://github.com/lgtmco/lgtm

**Comment by strk** [comment]
Github system documentation: https://help.github.com/articles/about-protected-branches/

**Comment by bradrydzewski** [comment]
I recommend reviewing the GitHub protected branches functionality. They have iterated on it quite significantly over the past 8 months and it is really shaping up nicely. Here are some quick highlights:

* any branch can be protected
* protected branches cannot be pushed to directly
  * unless the user account is part of a white list (configurable)
  * or unless you are a repository administrator
* protected branches can include required status checks
  * cannot merge until status checks are passing (ie CI checks, Coverage checks, Approvals)
  * administrator can override and force merge
* protected branches must be up-to-date with the target branch
  * merge button is disabled if it falls behind
  * administrator can override

The important thing here is that status checks unlock the protected branch so you can proceed with merging. This allows for much more custom and complex workflows for approving pull requests than something like simple voting.

I would advise against putting any approval algorithms in Gogs directly simply because, speaking from experience, it will only satisfy the needs of a subset of your users with others asking for more customization for more complex workflows. Using required status checks and delegating to 3rd party services is definitely the way to go IMO.

**Comment by Unknwon** [comment]
👍  for @bradrydzewski 's comment, but not quite understand this one:

> protected branches must be up-to-date with the target branch

**Comment by strk** [comment]
On Mon, Jul 18, 2016 at 03:05:32AM -0700, 无闻 wrote:
> > protected branches must be up-to-date with the target branch

I guess it means they are rebased to the target branch.
In other words, they can be merged with a fast-forward operation
(no need for a merge commit).

**Comment by Unknwon** [comment]
@strk no merge conflict or...?

**Comment by bradrydzewski** [comment]
@Unknwon hopefully this helps clarify  :)

> Please also note that when protected branches prevent you from merging a pull request that has fallen behind the target branch. That is to say, if a pull request is merged, all outstanding pull requests will need to re-sync before they can be merged.

Notice the message below "This branch is out-of-date with the base branch" indicating that I need to rebase with the base branch. This prevents the situation where, for example, you have PR 1 passing, PR 2 merged, and PR 2 has changes that break PR 1.

![7938e8f1-fda2-42ca-b876-c8bb313d9318](https://cloud.githubusercontent.com/assets/817538/16911708/e9dbbd18-4c94-11e6-81b8-8803f653162e.png)

**Comment by strk** [comment]
Independently of merge conflict, a fast-forward merge
is one from a branch which is based on the top of the
target branch.

Seek "fast-forward" here:
https://git-scm.com/book/en/v2/Git-Branching-Basic-Branching-and-Merging

**Comment by hrobertson** [comment]
@bradrydzewski Would this 'up-to-date' requirement be an optional setting per protected branch?

Some people may want the option to merge PRs that are behind, so long as there are no merge conflicts.

Assuming this was an option: For branches with this switched on, a new PR would have to be created?
For branches with it switched off, pending PRs should have checks (including CI) re-triggered/re-evaluated.

**Comment by bradrydzewski** [comment]
> Would this 'up-to-date' requirement be an optional setting per protected branch?

It is not an option with GitHub. If you are an administrator, however, you have the ability to override protected branches and merge as long as there are no conflicts. 

> Some people may want the option to merge PRs that are behind, so long as there are no merge conflicts.

I can't really argue one way or another since I don't have access to real world usage metrics or feedback. I can only reference the GitHub implementation which prevents merging out-of-date pull requests and does not make this optional.

**Comment by bradrydzewski** [comment]
ok, I take that back, github has added the option to toggle this feature. The feature is in beta and they are changing it on a weekly basis, but I am told it will be out of beta soon. So what you see in the screenshots and read in their official docs is probably pretty close to their final implementation

<img width="776" alt="screen shot 2016-07-18 at 5 27 16 am" src="https://cloud.githubusercontent.com/assets/817538/16914695/5e124176-4ca8-11e6-8fd1-74598bd91304.png">

**Comment by gilou** [comment]
There could also be a "simple" permission setup for branches... Deciding for on organization who can write to what branches... Allowing pull requests to be created as branches, but no write on master for example. A gitolite way of handling things can be a good start... as in: perms on org/user for the repos, perms for branches.. or allowing branch named dev/xxx for devs & so on...

**Comment by Unknwon** [comment]
Note: 

1. For detect and disable force push: http://stackoverflow.com/questions/10319110/how-to-detect-a-forced-update
2. Then Gogs needs to use both `update` and `pre-receive` hooks, and in order to allow user has custom scripts for these two hooks, we need to implement https://github.com/gogits/gogs/issues/1623 first.

**Comment by grotesque** [comment]
see my comments here: https://github.com/gogits/gogs/issues/3594
+1

**Comment by CharlesLR** [comment]
Hi
I develop a little script for protect a specific branch.
In Gogs, you past the content in the Git Hooks pre-receive : [https://github.com/CharlesLR/git-hook/blob/master/pre-receive.txt](url)
I hope that will help you, before a new version of Gogs

**Comment by denji** [comment]
`.git/hooks` is not a secure solution, it is still possible to rewrite hooks

**Comment by grotesque** [comment]
@CharlesLR 
Hi,
I am not able to view that url (page not found). Can you look into this?

**Comment by grotesque** [comment]
@denji 
Hi,
I don't understand how is it possible to rewrite hooks? This hook will be part of the server's .git folder. How can anybody rewrite them?

**Comment by denji** [comment]
@grotesque Destroy branch?

**Comment by CharlesLR** [comment]
@grotesque 
Link is OK now :)

**Comment by ElderJames** [comment]
需要这个功能！

**Comment by afdev82** [comment]
> Isn't it possible to implement simple protection (e.g. only allow some users to push to specific branches) via git-hooks on the server?
In order to do this we need to know the user that is pushing, I didn't find any doc about this.
How could I access the current user inside the hook?

**Comment by hrobertson** [comment]
@afdev82 The username is probably available in an environment variable. Write a hook to log the env vars so you know what's available.
Rather than replying here, please create a new issue regarding the need for documentation on the hooks environment.

Please let's keep this issue tidy. :smiley:

**Comment by afdev82** [comment]
@hrobertson Yes, you are right, but I didn't find any better place to write to get this information (I didn't want to open an issue only for that).
Anyway, thanks for the reply, I will try your suggestion.

**Comment by CihanSari** [comment]
Would this also include ability to hide source on some branches? I have seen some other issues which would address this need closed and referred to this one. Example scenario:

Software and hardware developers have access to the source code, production engineers / test engineers can open issues / provide feedback for next milestones. We don't want to make the code open for everyone, but issues should be open to at least a big part of the company. Currently we are making the Gogs completely hidden from the production and add a layer (somebody) to receive feedback and create the issues, ask for feedback and comment etc.

**Comment by strk** [comment]
Software, like information, shall be free

**Comment by Unknwon** [comment]
Initial version of protected branches is now on `develop` and online on demo site.

Since this thread is fairly long, **please file new issues for suggestions and bug reports**, thank you!

**Comment by fatbigbright** [comment]
Thanks for updates for protected branch feature in RC0.10. I wonder if there is a way to allow repository owner freely push code to the protected branch?

**Comment by Unknwon** [comment]
@fatbigbright It doesn't treat owner differently right now and I also think this should not be the default behavior. 

But can add always another option to allow owner freely do anything to the branch.

**Comment by fatbigbright** [comment]
@Unknwon  Sorry I am not quite sure about what you said about "add always another option", could you give more hints?

**Comment by hrobertson** [comment]
He means he will consider adding that as an option. I agree with him that
it should not be default behaviour.

**Comment by 502647092** [comment]
+1

**Comment by PHolker** [comment]
In gogs, you can (soft)protect a branch in the web interface. 
Still, any user may continue to push (NOT force-push, though)
or even perform pull-request merges via the web interface.

Remidy comes in the form of git hooks, as both the user name and branch can be accessed. 

Beware of the Gogs bug: git-hooks edited via web interface of gogs will be stored in .git/custom_hooks BUT will not work, so you must modify the .got/hooks/pre-receive file instead.

Here is my script to (hard)protect a branch, assuming you do not enable pull requests for it! 
What is does: limit pushes to a protected branch to a list of user names.

happy coding ...

[pre-receive-branch-protection.txt](https://github.com/gogits/gogs/files/1456474/pre-receive-branch-protection.txt)
